### PR TITLE
Unify and simplify TrN for i18n

### PR DESCRIPTION
**Pull request by wxiaoguang**
Refer: https://github.com/go-gitea/gitea/pull/18135#issuecomment-1003246099

Now we have a unique and simple `TrN`, and make the fix of PR #18135 also use the better `TrN` logic.

### CODE REVIEW

File: routers/web/repo/migrate.go [hunk 1]
```diff
@@ -81,13 +81,8 @@ func handleMigrateError(ctx *context.Context, owner *user_model.User, err error,
...
 		ctx.RenderWithErr(ctx.Tr("form.2fa_auth_required"), tpl, form)
 	case repo_model.IsErrReachLimitOfRepo(err):
-		var msg string
 		maxCreationLimit := ctx.User.MaxCreationLimit()
-		if maxCreationLimit == 1 {
-			msg = ctx.Tr("repo.form.reach_limit_of_creation_1", maxCreationLimit)
-		} else {
-			msg = ctx.Tr("repo.form.reach_limit_of_creation_n", maxCreationLimit)
-		}
+		msg := ctx.TrN(maxCreationLimit, "repo.form.reach_limit_of_creation_1", "repo.form.reach_limit_of_creation_1", maxCreationLimit)
```
Comment (KN4CK3R) [inline routers/web/repo/migrate.go]:
```suggestion
		msg := ctx.TrN(maxCreationLimit, "repo.form.reach_limit_of_creation_1", "repo.form.reach_limit_of_creation_n", maxCreationLimit)
```

File: routers/web/repo/repo.go [hunk 2]
```diff
@@ -162,13 +162,8 @@ func Create(ctx *context.Context) {
...
 	switch {
 	case repo_model.IsErrReachLimitOfRepo(err):
-		var msg string
 		maxCreationLimit := ctx.User.MaxCreationLimit()
-		if maxCreationLimit == 1 {
-			msg = ctx.Tr("repo.form.reach_limit_of_creation_1", maxCreationLimit)
-		} else {
-			msg = ctx.Tr("repo.form.reach_limit_of_creation_n", maxCreationLimit)
-		}
+		msg := ctx.TrN(maxCreationLimit, "repo.form.reach_limit_of_creation_1", "repo.form.reach_limit_of_creation_1", maxCreationLimit)
```
Comment (KN4CK3R) [inline routers/web/repo/repo.go]:
```suggestion
		msg := ctx.TrN(maxCreationLimit, "repo.form.reach_limit_of_creation_1", "repo.form.reach_limit_of_creation_n", maxCreationLimit)
```

File: routers/web/repo/setting.go [hunk 3]
```diff
@@ -610,11 +610,8 @@ func SettingsPost(ctx *context.Context) {
 
 		if !ctx.Repo.Owner.CanCreateRepo() {
 			maxCreationLimit := ctx.User.MaxCreationLimit()
-			if maxCreationLimit == 1 {
-				ctx.Flash.Error(ctx.Tr("repo.form.reach_limit_of_creation_1", maxCreationLimit))
-			} else {
-				ctx.Flash.Error(ctx.Tr("repo.form.reach_limit_of_creation_n", maxCreationLimit))
-			}
+			msg := ctx.TrN(maxCreationLimit, "repo.form.reach_limit_of_creation_1", "repo.form.reach_limit_of_creation_1", maxCreationLimit)
```
Comment (KN4CK3R) [inline routers/web/repo/setting.go]:
```suggestion
			msg := ctx.TrN(maxCreationLimit, "repo.form.reach_limit_of_creation_1", "repo.form.reach_limit_of_creation_n", maxCreationLimit)
```

File: routers/web/repo/migrate.go (hunk 1 above)
Comment (wxiaoguang) [reply routers/web/repo/migrate.go]: Oh sorry, my bad .....

**Review by zeripath — APPROVED** [review]

File: modules/translation/translation.go [hunk 4]
```diff
@@ -99,3 +100,66 @@ func (l *locale) Language() string {
 func (l *locale) Tr(format string, args ...interface{}) string {
 	return i18n.Tr(l.Lang, format, args...)
 }
+
+// Language specific rules for translating plural texts
+var trNLangRules = map[string]func(int64) int{
```
Comment (Gusted) [inline modules/translation/translation.go:105]: Off-Topic:

I didn't knew about this one, is this documented somewhere for translators?

File: routers/web/repo/repo.go [hunk 5]
```diff
@@ -162,13 +162,8 @@ func Create(ctx *context.Context) {
 func handleCreateError(ctx *context.Context, owner *user_model.User, err error, name string, tpl base.TplName, form interface{}) {
 	switch {
 	case repo_model.IsErrReachLimitOfRepo(err):
-		var msg string
 		maxCreationLimit := ctx.User.MaxCreationLimit()
```
Comment (Gusted) [inline routers/web/repo/repo.go]:
```suggestion
		maxCreationLimit := owner.MaxCreationLimit()
```
#18153

File: modules/translation/translation.go [hunk 6]
```diff
@@ -99,3 +100,66 @@ func (l *locale) Language() string {
...
+		c = int64(t)
+	} else if t, ok := cnt.(int16); ok {
+		c = int64(t)
+	} else if t, ok := cnt.(int32); ok {
+		c = int64(t)
+	} else if t, ok := cnt.(int64); ok {
+		c = t
+	} else {
+		return l.Tr(keyN, args...)
+	}
```
Comment (Gusted) [inline modules/translation/translation.go:155]:
```suggestion
	switch t := cnt.(type) {
	case int:
		c = int64(t)
	case int8:
		c = int64(t)
	case int16:
		c = int64(t)
	case int32:
		c = int64(t)
	case int64:
		c = t
	default:
		return l.Tr(keyN, args...)
	}
```

File: modules/csv/csv_test.go [hunk 7]
```diff
@@ -47,11 +57,11 @@ a,	b	c
 	e	f
 g	h	i
 j		l
-m	n,	
+m	n,\t
```
Comment (6543) [inline modules/csv/csv_test.go:58]: looks unrelated

File: modules/csv/csv_test.go [hunk 8]
```diff
@@ -21,14 +22,23 @@ func TestCreateReader(t *testing.T) {
 	assert.Equal(t, ',', rd.Comma)
 }
 
-//nolint
+func decodeSlashes(s string) string {
+	s = strings.ReplaceAll(s, "\n", "\\n")
+	s = strings.ReplaceAll(s, "\"", "\\\"")
+	decoded, err := strconv.Unquote(`"` + s + `"`)
+	if err != nil {
+		panic("unable to decode string:" + err.Error() + "\n" + s)
```
Comment (6543) [inline modules/csv/csv_test.go]: panic in test sounds like a bad idea!

File: modules/translation/translation.go [hunk 9]
```diff
@@ -99,3 +100,66 @@ func (l *locale) Language() string {
 func (l *locale) Tr(format string, args ...interface{}) string {
 	return i18n.Tr(l.Lang, format, args...)
 }
+
+// Language specific rules for translating plural texts
+var trNLangRules = map[string]func(int64) int{
+	"en-US": func(cnt int64) int {
```
Comment (6543) [inline modules/translation/translation.go:107]: de-DE is missing and it's same as "en-US"

File: modules/translation/translation.go (hunk 6 above)
Comment (6543) [reply modules/translation/translation.go:155]:
```suggestion
	switch t := cnt.(type) {
	case int, int8, int16, int32:
		c = int64(t)
	case int64:
	  c = t
	default:
		return l.Tr(keyN, args...)
	}
```

Comment (Gusted) [reply modules/translation/translation.go:155]: This won't work. go(without proper generics implementation) will now mark `t` on line 146 as `interface{}`(which cannot be converted to `int64`) because go doesn't have support for 1 variable to be multiple types.

Comment (6543) [reply modules/translation/translation.go:155]: uff 😅

File: modules/csv/csv_test.go (hunk 8 above)
Comment (wxiaoguang) [reply modules/csv/csv_test.go]: It's in unit test (for helper), so I think it's fine. Otherwise it would make code too complex.

File: modules/csv/csv_test.go (hunk 7 above)
Comment (wxiaoguang) [reply modules/csv/csv_test.go:58]: It fixes the lint and unit test, otherwise some IDE would trim the last space and make unit test fail.

The file `csv_test.go` was changed in this PR (TrN interface) so I have to fix other related code.

File: modules/translation/translation.go (hunk 4 above)
Comment (wxiaoguang) [reply modules/translation/translation.go:105]: I have no idea about `is this documented somewhere for translators`.

IMO it's mostly a knowledge for developers. If a developer write "text_1" and "text_n", the translators could just translate them without knowing `TrN`

File: modules/csv/csv_test.go (hunk 8 above)
Comment (6543) [reply modules/csv/csv_test.go]: just pass `t *testing.T` as argument & t.FailNow()

File: routers/web/repo/repo.go (hunk 5 above)
Comment (wxiaoguang) [reply routers/web/repo/repo.go]: Merged from main

File: modules/csv/csv_test.go (hunk 8 above)
Comment (wxiaoguang) [reply modules/csv/csv_test.go]: I prefer not, the `decodeString` is not part of the unit test ........ the panic should never happen

Comment (wxiaoguang) [reply modules/csv/csv_test.go]: Hmm, finally, I updated to `func decodeSlashes(t *testing.T, s string) string`

File: modules/translation/translation.go (hunk 9 above)
Comment (wxiaoguang) [reply modules/translation/translation.go:107]: It's default to `en-US` for most languages, so there won't be a problem (and no duplicated code)

image

Comment (wxiaoguang) [reply modules/translation/translation.go:107]: I added a comment here: `// the default rule is "en-US" if a language isn't listed here`, it's clear to everyone now 😊

**Review by 6543 — APPROVED** [review]

File: modules/csv/csv_test.go (hunk 8 above)
Comment (6543) [reply modules/csv/csv_test.go]: I just have seen to many `panic`s in legacy code that wen evil at some point as code changes

Comment (6543) [reply modules/csv/csv_test.go]: thanks :+1:
